ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,605 @@
+# Ruby - Phase 2
+
+This proposal is for implementing workflows in the Ruby SDK.
+
+See [phase 1](sdk-phase-1.md) for details on phase 1 which includes client and activities (granted some things may have
+changed since first authored).
+
+Notes for this proposal (like phase 1):
+
+* Everything is subject to change both during proposal time and afterwards during implementation.
+* For each place a decision is made, a "💭 Why X? ..." aside is present explaining why.
+* For each place an open question exists, a "❓ X?" exists to ask it.
+
+Here is a high-level overview of potentially controversial decisions:
+
+* Base class for activities is now `Temporalio::Activity::Definition` to be consistent with the base class
+* Class methods are dynamically made available for signals, queries, and updates to use from clients.
+  * So `MyWorkflow.my_signal` can be used in a client, and it just returns a signal definition, unlike the instance
+    form.
+* By default `TracePoint` will be enabled for workflow threads that will make sure users are not calling illegal calls,
+  but there is a way to disable for blocks of code.
+* `Ractor`s will be used for state isolation by default assuming decent performance but easy to disable and mostly
+  hidden from users.
+
+These points are explained below.
+
+## Workflow Definition/Usage
+
+Examples below are taken from
+[this Python sample](https://github.com/temporalio/samples-python/tree/main/message_passing/introduction), so not too
+many comments about general non-Ruby behavior are inline.
+
+Example definition:
+
+```ruby
+class CallGreetingServiceActivity < Temporalio::Activity::Definition
+  GREETINGS = {
+    arabic: 'مرحبا بالعالم',
+    chinese: '你好，世界',
+    english: 'Hello, world',
+    french: 'Bonjour, monde',
+    hindi: 'नमस्ते दुनिया',
+    portuguese: 'Olá mundo',
+    spanish: 'Hola mundo'
+  }
+
+  def execute(language)
+    # Simulate a network call
+    sleep(0.2)
+    GREETINGS[language.to_sym]
+  end
+end
+
+class GreetingWorkflow < Temporalio::Workflow::Definition
+  workflow_query_attr :language
+
+  def initialize(name)
+    @greetings = {
+      chinese: '你好，世界',
+      english: 'Hello, world'
+    }
+    @language = :english
+    @mutex = Temporalio::Workflow::Mutex.new
+  end
+
+  def execute
+    Temporalio::Workflow.wait_condition do
+      @approved_for_release && Temporalio::Workflow.all_handlers_finished?
+    end
+    @greetings[@language]
+  end
+
+  workflow_query
+  def languages(input)
+    return CallGreetingServiceActivity::GREETINGS.keys.sort if input[:include_unsupported]
+    @greetings.keys.sort
+  end
+
+  workflow_signal
+  def approve(input)
+    @approved_for_release = true
+    @approver_name = input[:name]
+  end
+
+  workflow_update
+  def set_language(language)
+    previous, @language = @language, language.to_sym
+    previous
+  end
+
+  workflow_update_validator(:set_language)
+  def validate_set_language(language)
```

Comment:
What would be the experience if they don't keep the args in sync b/w validator and handler?  I'd like them to get reasonable feedback.
suggestion: This seems like one to see if sorbet/rbs can wrangle static typesafety via a generic, perhaps by putting in a nested class.

Reply:
This fails at register time like every other register-time validator in every other SDK